File: clojure/Day06.clj
(ns Day06)

; Returns the number of orbits the given object is contained in (i.e. direct orbit + indirect orbits)
(defn orbit-centers [orbit-map object]
  (take-while some? (drop 1 (iterate orbit-map object))))

; Returns the sum of the orbit centers for each object in the orbit map.
(defn direct-and-indirect-orbits [orbit-map]
  (apply + (map count (map (partial orbit-centers orbit-map) (keys orbit-map)))))

; Counts the amount of elements in coll before element appears. Returns (count coll) if it doesn't appear at all.
(defn count-until [element coll]
  (count (take-while (partial not= element) coll)))

; Calculates the minimum amount of traversals required in an orbit map to move from object to destination
(defn traversals [orbit-map object destination]
  (let [object-centers (orbit-centers orbit-map object)
        destination-centers (orbit-centers orbit-map destination)
        first-common-center (some (set destination-centers) object-centers)]
    (+ (count-until first-common-center object-centers)
       (count-until first-common-center destination-centers))))

; Parses the input format into a map of object -> orbit center
(defn parse-orbit-map [raw]
  (apply hash-map (flatten (map (comp reverse #(.split #"\)" %)) (.split #"\n" raw)))))

(def input (parse-orbit-map (slurp (first *command-line-args*))))
(println "Total number of (in)direct orbits:" (direct-and-indirect-orbits input))
(println "Traversals required to get from you to santa:" (traversals input "YOU" "SAN"))

File: haskell/Day06.hs
module Main where

import Common

-- Edge with start and end vertex
data Edge t = Edge t t deriving (Show, Eq)

-- Count all (transitive) edges that ends in the end vertex
count :: (Eq t) => [Edge t] -> t -> Int
count edges end = helper edges edges end
  where
    helper total [] end = 0
    helper total ((Edge a b):xs) end
      | (b == end) =  1 + (helper total total a) + (helper total xs end)
      | otherwise = helper total xs end

points :: (Eq t) => [Edge t] -> t -> t -> [t]
points edges start end = helper edges edges start end
  where
    helper total [] start end = []
    helper total ((Edge a b):xs) start end
      | (a == start && b == end) = [a, b]
      | (b == end) =  [b] ++ (helper total total start a) ++ (helper total xs start end)
      | otherwise = helper total xs start end

intersect :: [Edge String] -> [String]
intersect edges = [a | a <- points edges "COM" "YOU", elem a (points edges "COM" "SAN")]

distance :: (Eq t) => [Edge t] -> t -> t -> Int
distance edges start end = length (points edges start end) - 1


minPoint :: (Ord t) => [Edge t] -> [t] -> t -> t
minPoint edges vertexes center = snd $ minimum $ map (\x ->  (distance edges x center, x)) vertexes


{- Handle input and solve executions -}

-- Solve first challenge
solve1 :: Input -> Int
solve1 (Lines (Just lines)) = sum (map (count edges) targets)
  where
    targets = map (\(Edge a b) -> b) edges
    edges = map parseEdge lines
    parseEdge line = Edge ((split line ')') !! 0) ((split line ')') !! 1)

-- Solve second challenge
solve2 :: Input -> Int
solve2 (Lines (Just lines)) = (distance edges (minP "YOU") "YOU") + (distance edges (minP "SAN") "SAN") - 2
  where
    minP = minPoint edges (intersect edges)
    targets = map (\(Edge a b) -> b) edges
    edges = map parseEdge lines
    parseEdge line = Edge ((split line ')') !! 0) ((split line ')') !! 1)

-- Print challenge result
main = do
  solve "Number of (in-)direct orbits" (Lines Nothing) (solve1)
  solve "Minimum number of orbital transfers" (Lines Nothing) (solve2)